File: src/lib/scala/Slate.Common/src/main/scala/slate/common/envs/Env.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common.envs

import slate.common.{Strings}


/**
  * Represents a system environment
  * @param name   : e.g. Quality Assurance
  * @param mode   : e.g. Qa
  * @param region : New york
  * @param desc   : Qa environment 1 in new york
  */
case class Env( name:String, mode:EnvMode, region:String = "", desc:String = "" ) extends EnvSupport {

  /**
    * "qa1:qa"
    * @return
    */
  def key : String = name + ":" + mode.name


  override def isEnv(envMode: String):Boolean = mode.name == envMode


  override def isEnv(envMode: EnvMode):Boolean = mode == envMode
}


object Env {

  /**
   * List of default environments supported in slate kit
   * @return
   */
  def defaults(): Envs = {
    val all = List[Env](
      Env("loc", Dev , desc = "Dev environment (local)" ),
      Env("dev", Dev , desc = "Dev environment (shared)" ),
      Env("qa1", Qa  , desc = "QA environment  (current release)" ),
      Env("qa2", Qa  , desc = "QA environment  (last release)" ),
      Env("stg", Uat , desc = "STG environment (demo)" ),
      Env("pro", Prod, desc = "LIVE environment" )
    )
    new Envs(all, Some(all(0)))
  }


  /**
   * parses the environment name e.g. "qa1:qa" = name:mode
   * @param env
   * @return
   */
  def parse(env:String):Env = {
    val tokens = Strings.split(env, ':')

    // e.g. "dev1", "dev", dev1:dev")
    if(tokens.length == 1)
      Env(tokens(0), interpret(tokens(0)))
    else
      Env(tokens(0), interpret(tokens(1)))
  }


  /**
    * interprets the string representation of an environment into the type object
    * @param mode
    * @return
    */
  def interpret(mode:String):EnvMode = {
    val m = Option(mode).getOrElse("").toLowerCase
    m match {
      case Dev.name  => Dev
      case Qa.name   => Qa
      case Uat.name  => Uat
      case Prod.name => Prod
      case Dis.name  => Dis
      case _         => Other(mode)
    }
  }

}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/envs/Env.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.envs

/**
 * Represents a system environment
 * @param name : e.g. Quality Assurance test
 * @param mode : e.g. Qat
 * @param region : new york
 * @param desc : Qat environment 1 in new york
 */
data class Env(
    @JvmField
    val name: String,

    @JvmField
    val mode: EnvMode,

    @JvmField
    val region: String = "",

    @JvmField
    val desc: String = ""
) : EnvSupport {

    /**
     * "qa1:qa"
     * @return
     */
    val key: String get() = name + ":" + mode.name

    override fun isEnvName(envMode: String): Boolean = mode.name == envMode

    override fun isEnv(envMode: EnvMode): Boolean = mode == envMode

    companion object {

        @JvmStatic
        val empty = Env("", EnvMode.Dev, "", "")

        /**
         * The list of defaults environments to choose from.
         * An environment definition is defined by its name, mode
         * The key is built up from name and mode as {name}.{mode}
         * e.g. "qa1.QA"
         *
         * Each of these environments should map to an associated env.{name}.conf
         * config file in the /resources/ directory. But there is no dependency
         * on this Env component to a Config component
         *
         * e.g.
         * /resources/env.conf     ( common      config )
         * /resources/env.loc.conf ( local       config )
         * /resources/env.dev.conf ( development config )
         * /resources/env.qa1.conf ( qa1         config )
         * /resources/env.qa2.conf ( qa2         config )
         * /resources/env.stg.conf ( staging     config )
         * /resources/env.pro.conf ( production  config )
         *
         * @return
         */
        @JvmStatic
        fun defaults(): Envs =
            Envs(listOf(
                Env("loc", EnvMode.Dev, desc = "Dev environment (local)"),
                Env("dev", EnvMode.Dev, desc = "Dev environment (shared)"),
                Env("qa1", EnvMode.Qat, desc = "QA environment  (current release)"),
                Env("qa2", EnvMode.Qat, desc = "QA environment  (last release)"),
                Env("stg", EnvMode.Uat, desc = "STG environment (demo)"),
                Env("pro", EnvMode.Pro, desc = "LIVE environment")
            ))

        /**
         * parses the environment name e.g. "qa1:qa" = name:mode
         * @param env
         * @return
         */
        @JvmStatic
        fun parse(text: String): Env {
            if (text == "''" || text == "\"\"") {
                return Env("", EnvMode.Dev, desc = "Default environment")
            }
            val tokens = text.split(':')

            // e.g. "dev1", "dev", dev1:dev")
            val env = if (tokens.size == 1)
                Env(tokens[0], interpret(tokens[0]))
            else
                Env(tokens[0], interpret(tokens[1]))

            return env
        }

        /**
         * interprets the string representation of an environment into the type object
         * @param mode
         * @return
         */
        @JvmStatic
        fun interpret(mode: String): EnvMode =
                when (mode) {
                    EnvMode.Dev.name -> EnvMode.Dev
                    EnvMode.Qat.name -> EnvMode.Qat
                    EnvMode.Uat.name -> EnvMode.Uat
                    EnvMode.Pro.name -> EnvMode.Pro
                    EnvMode.Dis.name -> EnvMode.Dis
                    else -> EnvMode.Other(mode)
                }
    }
}
